kong/clustering/wrpc_data_plane.lua

```diff
@@ -0,0 +1,379 @@
+        peer.config_obj.next_config_version = tonumber(data.version)
+        if peer.config_semaphore:count() <= 0 then
+          -- the following line always executes immediately after the `if` check
+          -- because `:count` will never yield, end result is that the semaphore
+          -- count is guaranteed to not exceed 1
+          peer.config_semaphore:post()
+        end
+      end
+      return { accepted = true }
+    end)
+  end
+
+  return wrpc_config_service
+end
+
+function _M:communicate(premature)
+
+  if premature then
+    -- worker wants to exit
+    return
+  end
+
+  local conf = self.conf
+
+  -- TODO: pick one random CP
+  local address = conf.cluster_control_plane
+  local log_suffix = " [" .. address .. "]"
+
+  local c = assert(ws_client:new(WS_OPTS))
+  local uri = "wss://" .. address .. "/v1/outlet?node_id=" ..
+              kong.node.get_id() ..
+              "&node_hostname=" .. kong.node.get_hostname() ..
+              "&node_version=" .. KONG_VERSION
+
+  local opts = {
+    ssl_verify = true,
+    client_cert = self.cert,
+    client_priv_key = self.cert_key,
+    protocols = "wrpc.konghq.com",
+  }
+  if conf.cluster_mtls == "shared" then
+    opts.server_name = "kong_clustering"
+  else
+    -- server_name will be set to the host if it is not explicitly defined here
+    if conf.cluster_server_name ~= "" then
+      opts.server_name = conf.cluster_server_name
+    end
+  end
+
+  local reconnection_delay = math.random(5, 10)
+  do
+    local res, err = c:connect(uri, opts)
+    if not res then
+      ngx_log(ngx_ERR, _log_prefix, "connection to control plane ", uri, " broken: ", err,
+        " (retrying after ", reconnection_delay, " seconds)", log_suffix)
+
+      assert(ngx.timer.at(reconnection_delay, function(premature)
+        self:communicate(premature)
+      end))
+      return
+    end
+  end
+
+  local config_semaphore = semaphore.new(0)
+  local peer = wrpc.new_peer(c, get_config_service(), { channel = true })
+
+  peer.config_semaphore = config_semaphore
+  peer.config_obj = self
+  peer:spawn_threads()
+
+  do
+    local resp, err = peer:call_wait("ConfigService.ReportMetadata", { plugins = self.plugins_list })
+    if type(resp) == "table" then
+      err = err or resp.error
+      resp = resp[1] or resp.ok
+    end
+    if type(resp) == "table" then
+      resp = resp.ok or resp
+    end
+
+    if not resp then
+      ngx_log(ngx_ERR, _log_prefix, "Couldn't report basic info to CP: ", err)
+      assert(ngx.timer.at(reconnection_delay, function(premature)
+        self:communicate(premature)
+      end))
+    end
+  end
+
+  -- Here we spawn two threads:
+  --
+  -- * config_thread: it grabs a received declarative config and apply it
+  --                  locally. In addition, this thread also persists the
+  --                  config onto the local file system
+  -- * ping_thread: performs a ConfigService.PingCP call periodically.
+
+  local ping_immediately
```

Comment:
`ping_immediately` is never set in this file, while this field is set in `data_plane.lua`.

Reply:
good catch.  in the original `data_plane.lua` it's used to send the first ping with the config hash and prime the config pushing from the control_plane; but in wrpc the config is initiated by the control plane and this isn't required.

i'll remove the variable